### ISSUE TITLE
getBitrateEstimate() works something wrong.

### ISSUE THREAD

**Issue by hereisclient** [google/ExoPlayer#5021]
### Issue description
i'm developing my app using Exoplayer.
most of all. it worked fine.
but i want to show current bitrate on my player information.
So i used DefaultBandwidthMeter.
and i checked bitrate every 10 sec.

DefaultBandwidthMeter BANDWIDTH_METER = new DefaultBandwidthMeter();

class BitrateTask extends TimerTask {
    @Override
    public void run() {
      Log.d("COT", "send bitrate=" + BANDWIDTH_METER.getBitrateEstimate());
    }
  }

if it's not issue,  that times,  could you tell me how to get bitrate in this sample url?

### Reproduction steps
always 

### Link to test content
i will send you test url with issue number.

### Version of ExoPlayer being used
2.8.3

### Device(s) and version(s) of Android being used
device name : shaomi mibox 3
android version : 8.0

### A full bug report captured from the device

2018-10-30 13:52:27.596 12865-12865/com.google.android.exoplayer2.demo D/EventLogger: downstreamFormatChanged [10.87, 9.81, window=0, period=0, id=5, mimeType=video/avc, bitrate=3000000, res=1280x720]
2018-10-30 13:52:28.577 12865-13088/com.google.android.exoplayer2.demo D/COT: send bitrate=34817456
2018-10-30 13:52:31.577 12865-13088/com.google.android.exoplayer2.demo D/COT: send bitrate=34817456
**2018-10-30 13:52:32.467 12865-13075/com.google.android.exoplayer2.demo I/MediaCodecLogger: com.google.android.exoplayer2.demo.HW.video.avc.bitrateInKbps = 2584**
2018-10-30 13:52:32.618 12865-13083/com.google.android.exoplayer2.demo I/MediaCodecLogger: com.google.android.exoplayer2.demo.SW.audio.mp4a.bitrateInKbps = 95
**2018-10-30 13:52:34.577 12865-13088/com.google.android.exoplayer2.demo D/COT: send bitrate=34817456**
2018-10-30 13:52:37.485 12865-13075/com.google.android.exoplayer2.demo I/MediaCodecLogger: com.google.android.exoplayer2.demo.HW.video.avc.bitrateInKbps = 2039
2018-10-30 13:52:37.578 12865-13088/com.google.android.exoplayer2.demo D/COT: send bitrate=34817456
2018-10-30 13:52:37.709 12865-13083/com.google.android.exoplayer2.demo I/MediaCodecLogger: com.google.android.exoplayer2.demo.SW.audio.mp4a.bitrateInKbps = 94
2018-10-30 13:52:40.577 12865-13088/com.google.android.exoplayer2.demo D/COT: send bitrate=34817456
2018-10-30 13:52:42.600 12865-13075/com.google.android.exoplayer2.demo I/MediaCodecLogger: com.google.android.exoplayer2.demo.HW.video.avc.bitrateInKbps = 2363
2018-10-30 13:52:42.782 12865-13083/com.google.android.exoplayer2.demo I/MediaCodecLogger: com.google.android.exoplayer2.demo.SW.audio.mp4a.bitrateInKbps = 96
2018-10-30 13:52:43.577 12865-13088/com.google.android.exoplayer2.demo D/COT: send bitrate=34817456
2018-10-30 13:52:46.577 12865-13088/com.google.android.exoplayer2.demo D/COT: send bitrate=34817456
2018-10-30 13:52:47.720 12865-13075/com.google.android.exoplayer2.demo I/MediaCodecLogger: com.google.android.exoplayer2.demo.HW.video.avc.bitrateInKbps = 3075
2018-10-30 13:52:47.851 12865-13083/com.google.android.exoplayer2.demo I/MediaCodecLogger: com.google.android.exoplayer2.demo.SW.audio.mp4a.bitrateInKbps = 95
2018-10-30 13:52:49.577 12865-13088/com.google.android.exoplayer2.demo D/COT: send bitrate=34817456
2018-10-30 13:52:52.578 12865-13088/com.google.android.exoplayer2.demo D/COT: send bitrate=34817456
2018-10-30 13:52:52.837 12865-13075/com.google.android.exoplayer2.demo I/MediaCodecLogger: com.google.android.exoplayer2.demo.HW.video.avc.bitrateInKbps = 1421
2018-10-30 13:52:52.941 12865-13083/com.google.android.exoplayer2.demo I/MediaCodecLogger: com.google.android.exoplayer2.demo.SW.audio.mp4a.bitrateInKbps = 95
2018-10-30 13:52:55.579 12865-13088/com.google.android.exoplayer2.demo D/COT: send bitrate=34817456

**Comment by tonihei** [comment]
I think there is some confusion around the various bitrates here. 

The logs look correct and sensible:
 - There is a video track with a variable bitrate around 3 Mbps ("video.avc.bitrateInKbps = 2584")
 - There is also an audio track with a bitrate of 95 kbps ("audio.mp4a.bitrateInKbps = 95")
 - And your bandwidth meter measured a network bitrate of around 34 Mbps ("send bitrate=34817456")

If that wasn't the question, please rephrase a bit to let us know where you see a problem.

**Comment by hereisclient** [comment]
i sent you test vod url serveral hours ago.
please could you test my url that i sent?

if bandwidth value from getBitrateEstimate() is correct or not.

**Comment by tonihei** [comment]
Checked your stream but everything works as I explained above.

**Comment by hereisclient** [comment]
i just want to know  how i can estimate bitrate correctly.
you mean there is no solution to get right bitrate from test url by now? 
i really want know how to get right bitrate from exoplayer.
but actually , getBitrateEstimate() function doesn't work well.

if you know the other way to get bitrate from exoplayer. please let me know about that.
this is a little important issue for me.

**Comment by tonihei** [comment]
Which of the bitrates do you want to know? 
- If it's the network bitrate, `BandwidthMeter.getBitrateEstimate()` gives the best estimate you can get from the player. 
- If it's the average bitrate of the video or audio format, please listen to `AnalyticsListener.onDownstreamFormatChanged` to get the currently selected format. The Format contains a `bitrate` field which gives you the average bitrate of the format.
- For more fine-grained information, you can also listen to `AnalyticsListener.onLoadStarted` and similar callbacks to get information about the individual media chunks being loaded.

**Comment by sridhard** [comment]
@tonihei 

I also face issues in bitrate/bandwidth. I have implemented the onBandwidthSample of DefaultBandwidthMeter. The elapsedMs in this callback is coming very low. Actual network speed is 2-3 Mbps. But the bitrate in  onBandwidthSample is 20-30Mbps.

I have checked okHttpDataSource.java of exoplayer. In that I find that exoplayer is downloading the HTTP media synchronously as below:

@Override
  public long open(DataSpec dataSpec) throws HttpDataSourceException {
    this.dataSpec = dataSpec;
    this.bytesRead = 0;
    this.bytesSkipped = 0;
    transferInitializing(dataSpec);

    Request request = makeRequest(dataSpec);
    Response response;
    ResponseBody responseBody;
    try {
      **_this.response = callFactory.newCall(request).execute();
      response = this.response;_**
      responseBody = Assertions.checkNotNull(response.body());
      responseByteStream = responseBody.byteStream();
    } catch (IOException e) {
      throw new HttpDataSourceException(
          "Unable to connect to " + dataSpec.uri, e, dataSpec, HttpDataSourceException.TYPE_OPEN);
    }

After the above codeblock okHttpDataSource is giving the transferStarted() indication.

This means that the startTime is captured after the execute() is success which may be after downloading the content. That is the reason for very low elapsedMs value. I am getting this value in the range of 10-50ms.

**Comment by tonihei** [comment]
@sridhard That may be an issue and we were planning to look at the accuracy of our estimates in the near future. But I think it's unrelated to original question above.

**Comment by hereisclient** [comment]
oh i really got it what you're trying to say .

There is a lot of bitrate.

but my need is to get real bitrate during exoplay is playing somecontent not network bandwidth.
my goal is just like below log.

**2018-10-30 13:52:52.837 12865-13075/com.google.android.exoplayer2.demo I/MediaCodecLogger: com.google.android.exoplayer2.demo.HW.video.avc.bitrateInKbps = 1421
2018-10-30 13:52:52.941 12865-13083/com.google.android.exoplayer2.demo I/MediaCodecLogger: com.google.android.exoplayer2.demo.SW.audio.mp4a.bitrateInKbps = 95**

because smooth streaming bitrate is changing every time .
it depands on real network bandwidth.

If it's the average bitrate of the video or audio format, please listen to AnalyticsListener.onDownstreamFormatChanged to get the currently selected format. The Format contains a bitrate field which gives you the average bitrate of the format.
For more fine-grained information, you can also listen to AnalyticsListener.onLoadStarted and similar callbacks to get information about the individual media chunks being loaded.

but two case of your examples , as far as i know,  i can use  mp4 format.   
because mp4 contents has only  static bitrate.
so in conclusion, 
my question is 
how can i get bitrate (just like mediacodec log) using adaptive streaming?

**Comment by tonihei** [comment]
As I said above, for adaptive streaming, please listen to AnalyticsListener.onLoadStarted and/or AnalyticsListener.onLoadFinished to get information about the individual loaded chunks. This information includes the size and the media times of the chunk, so that you can calculate the bitrate from that.

**Comment by tonihei** [comment]
I assume the question was answered. Closing the issue, but feel free to reopen if needed.
